File: [A]tree/[A]tree-dfs-postorder/236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def dfs(node):
            if not node:
                return None
            left_cand = dfs(node.left)
            right_cand = dfs(node.right)
            if node.val == p.val or node.val == q.val:
                return node
            if left_cand and right_cand:
                return node
            if left_cand or right_cand:
                return left_cand or right_cand
            return None
        
        return dfs(root)
    
# time O(n)
# space O(n)
# using tree and dfs (postorder and recursive)
'''
1. dfs postorder (bottom up approach), helps to get the potential LCA info
'''

File: [A]tree/[A]tree-dfs-postorder/236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return dfs(root, p, q);
    }

    public TreeNode dfs(TreeNode node, TreeNode p, TreeNode q) {
        if (node == null) {
            return node;
        }
        TreeNode leftCand = dfs(node.left, p, q);
        TreeNode rightCand = dfs(node.right, p, q);
        if (leftCand != null && rightCand != null) {
            return node;
        }
        if (node.val == p.val || node.val == q.val) {
            return node;
        }
        if (leftCand != null) {
            return leftCand;
        }
        if (rightCand != null) {
            return rightCand;
        }
        return null;
    }
}

// time O(n)
// space O(n)
// using tree and dfs (postorder and recursive)
/*
1. dfs postorder (bottom up approach), helps to get the potential LCA info
*/